File: skills/english-cefr-level-rewriter/scripts/validate_cefr_readability.rb
# ABOUTME: Validates English text grammar complexity against a CEFR level using heuristic rules
require "optparse"

CEFR_ORDER = %w[A1 A2 B1 B2 C1 C2].freeze unless defined?(CEFR_ORDER)

PROFILES = {
  "A1" => {
    max_words_per_sentence: 10,
    max_clause_depth:       0,
    passive_allowed:        false,
    allowed_modals:         %w[can can't cannot].to_set,
  },
  "A2" => {
    max_words_per_sentence: 15,
    max_clause_depth:       1,
    passive_allowed:        false,
    allowed_modals:         %w[can can't cannot could will won't would].to_set,
  },
  "B1" => {
    max_words_per_sentence: 20,
    max_clause_depth:       1,
    passive_allowed:        true,
    allowed_modals:         %w[can can't cannot could will won't would must should may might shall].to_set,
  },
  "B2" => {
    max_words_per_sentence: 25,
    max_clause_depth:       2,
    passive_allowed:        true,
    allowed_modals:         %w[can can't cannot could will won't would must should may might shall need ought].to_set,
  },
  "C1" => { max_words_per_sentence: 35,  max_clause_depth: 3,   passive_allowed: true, allowed_modals: nil },
  "C2" => { max_words_per_sentence: 999, max_clause_depth: 999, passive_allowed: true, allowed_modals: nil },
}.freeze

SUBORDINATORS = /\b(that|which|who|whom|whose|when|where|while|because|although|since|unless|until|if|though|so that|in order that)\b/i
PASSIVE       = /\b(am|is|are|was|were|be|been|being)\s+(\w+ed|\w+en)\b/i
ALL_MODALS    = %w[can could will would must should may might shall need ought can't cannot won't].freeze

def split_sentences(text)
  text.strip.split(/(?<=[.!?])\s+/).map(&:strip).reject(&:empty?)
end

def count_words(sentence)
  sentence.scan(/\b\w+\b/).size
end

def count_subordinate_clauses(sentence)
  sentence.scan(SUBORDINATORS).size
end

def has_passive?(sentence)
  PASSIVE.match?(sentence)
end

def check_modals(sentence, allowed)
  return [] if allowed.nil?
  pattern = /\b(#{ALL_MODALS.map { |m| Regexp.escape(m) }.join("|")})\b/i
  sentence.scan(pattern).flatten.reject { |m| allowed.include?(m.downcase) }
end

def validate_readability(text, target_level)
  profile    = PROFILES[target_level.upcase]
  violations = []
  split_sentences(text).each do |sentence|
    snippet = sentence[0, 80]
    wc = count_words(sentence)
    if wc > profile[:max_words_per_sentence]
      violations << { sentence: snippet, rule: "sentence_length",
                      detail: "#{wc} words (max #{profile[:max_words_per_sentence]})" }
    end
    depth = count_subordinate_clauses(sentence)
    if depth > profile[:max_clause_depth]
      violations << { sentence: snippet, rule: "clause_depth",
                      detail: "#{depth} subordinate clause(s) (max #{profile[:max_clause_depth]})" }
    end
    if !profile[:passive_allowed] && has_passive?(sentence)
      violations << { sentence: snippet, rule: "passive_voice",
                      detail: "passive construction not allowed at this level" }
    end
    check_modals(sentence, profile[:allowed_modals]).each do |modal|
      violations << { sentence: snippet, rule: "modal_verb",
                      detail: "'#{modal}' not in allowed modal inventory for #{target_level.upcase}" }
    end
  end
  violations
end

def print_violations(violations)
  if violations.empty?
    puts "No violations found."
    return
  end
  violations.each do |v|
    puts "[#{v[:rule]}] #{v[:detail]}"
    puts "  → #{v[:sentence]}"
  end
end

if __FILE__ == $0
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: validate_cefr_readability.rb --text TEXT --level LEVEL"
    opts.on("--text TEXT",  "Text to validate")                { |v| options[:text]  = v }
    opts.on("--level LEVEL", CEFR_ORDER, "Target CEFR level") { |v| options[:level] = v }
  end.parse!

  abort("--text is required")  unless options[:text]
  abort("--level is required") unless options[:level]

  violations = validate_readability(options[:text], options[:level])
  print_violations(violations)
  exit(violations.empty? ? 0 : 1)
end

File: skills/english-cefr-level-rewriter/scripts/validate_cefr_readability.py
# ABOUTME: Validates English text grammar complexity against a CEFR level using heuristic rules
import argparse
import re
import sys

CEFR_ORDER = ["A1", "A2", "B1", "B2", "C1", "C2"]

PROFILES = {
    "A1": {
        "max_words_per_sentence": 10,
        "max_clause_depth": 0,
        "passive_allowed": False,
        "allowed_modals": {"can", "can't", "cannot"},
    },
    "A2": {
        "max_words_per_sentence": 15,
        "max_clause_depth": 1,
        "passive_allowed": False,
        "allowed_modals": {"can", "can't", "cannot", "could", "will", "won't", "would"},
    },
    "B1": {
        "max_words_per_sentence": 20,
        "max_clause_depth": 1,
        "passive_allowed": True,
        "allowed_modals": {
            "can", "can't", "cannot", "could", "will", "won't", "would",
            "must", "should", "may", "might", "shall",
        },
    },
    "B2": {
        "max_words_per_sentence": 25,
        "max_clause_depth": 2,
        "passive_allowed": True,
        "allowed_modals": {
            "can", "can't", "cannot", "could", "will", "won't", "would",
            "must", "should", "may", "might", "shall", "need", "ought",
        },
    },
    "C1": {
        "max_words_per_sentence": 35,
        "max_clause_depth": 3,
        "passive_allowed": True,
        "allowed_modals": None,
    },
    "C2": {
        "max_words_per_sentence": 999,
        "max_clause_depth": 999,
        "passive_allowed": True,
        "allowed_modals": None,
    },
}

_SUBORDINATORS = re.compile(
    r"\b(that|which|who|whom|whose|when|where|while|because|although|"
    r"since|unless|until|if|though|so that|in order that)\b",
    re.IGNORECASE,
)
_PASSIVE = re.compile(
    r"\b(am|is|are|was|were|be|been|being)\s+(\w+ed|\w+en)\b", re.IGNORECASE
)
_ALL_MODALS = {
    "can", "could", "will", "would", "must", "should", "may", "might",
    "shall", "need", "ought", "can't", "cannot", "won't",
}


def split_sentences(text: str) -> list:
    return [s.strip() for s in re.split(r"(?<=[.!?])\s+", text.strip()) if s.strip()]


def count_words(sentence: str) -> int:
    return len(re.findall(r"\b\w+\b", sentence))


def count_subordinate_clauses(sentence: str) -> int:
    return len(_SUBORDINATORS.findall(sentence))


def has_passive(sentence: str) -> bool:
    return bool(_PASSIVE.search(sentence))


def check_modals(sentence: str, allowed) -> list:
    if allowed is None:
        return []
    pattern = r"\b(" + "|".join(re.escape(m) for m in _ALL_MODALS) + r")\b"
    found = re.findall(pattern, sentence, re.IGNORECASE)
    return [m for m in found if m.lower() not in allowed]


def validate_readability(text: str, target_level: str) -> list:
    profile = PROFILES[target_level.upper()]
    violations = []
    for sentence in split_sentences(text):
        word_count = count_words(sentence)
        if word_count > profile["max_words_per_sentence"]:
            violations.append({
                "sentence": sentence[:80],
                "rule": "sentence_length",
                "detail": f"{word_count} words (max {profile['max_words_per_sentence']})",
            })
        depth = count_subordinate_clauses(sentence)
        if depth > profile["max_clause_depth"]:
            violations.append({
                "sentence": sentence[:80],
                "rule": "clause_depth",
                "detail": f"{depth} subordinate clause(s) (max {profile['max_clause_depth']})",
            })
        if not profile["passive_allowed"] and has_passive(sentence):
            violations.append({
                "sentence": sentence[:80],
                "rule": "passive_voice",
                "detail": "passive construction not allowed at this level",
            })
        for modal in check_modals(sentence, profile["allowed_modals"]):
            violations.append({
                "sentence": sentence[:80],
                "rule": "modal_verb",
                "detail": f"'{modal}' not in allowed modal inventory for {target_level.upper()}",
            })
    return violations


def _print_violations(violations: list) -> None:
    if not violations:
        print("No violations found.")
        return
    for v in violations:
        print(f"[{v['rule']}] {v['detail']}")
        print(f"  → {v['sentence']}")


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate CEFR grammar/readability level")
    parser.add_argument("--text", required=True, help="Text to validate")
    parser.add_argument("--level", required=True, choices=CEFR_ORDER, help="Target CEFR level")
    args = parser.parse_args()

    violations = validate_readability(args.text, args.level)
    _print_violations(violations)
    return 1 if violations else 0


if __name__ == "__main__":
    sys.exit(main())
